metadata-service/factories/src/main/java/com/linkedin/metadata/boot/steps/IngestRetentionPoliciesStep.java

```diff
@@ -0,0 +1,144 @@
+package com.linkedin.metadata.boot.steps;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
+import com.linkedin.metadata.boot.BootstrapStep;
+import com.linkedin.metadata.dao.utils.RecordUtils;
+import com.linkedin.metadata.entity.RetentionService;
+import com.linkedin.metadata.key.DataHubRetentionKey;
+import com.linkedin.retention.DataHubRetentionInfo;
+import java.io.File;
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.Nonnull;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.core.io.ClassPathResource;
+
+
+@Slf4j
+@RequiredArgsConstructor
+public class IngestRetentionPoliciesStep implements BootstrapStep {
+
+  private final RetentionService _retentionService;
+  private final boolean _disableRetention;
+  private final String pluginPath;
+
+  private static final ObjectMapper YAML_MAPPER = new ObjectMapper(new YAMLFactory());
+
+  @Nonnull
+  @Override
+  public ExecutionMode getExecutionMode() {
+    return ExecutionMode.ASYNC;
+  }
+
+  @Override
+  public String name() {
+    return "IngestRetentionPoliciesStep";
+  }
+
+  @Override
+  public void execute() throws IOException, URISyntaxException {
+    // 0. Execute preflight check to see whether we need to ingest policies
+    log.info("Ingesting default retention...");
+
+    // Whether we are at clean boot or not.
+    if (_disableRetention) {
+      log.info("IngestRetentionPolicies disabled. Skipping.");
+      return;
+    }
+
+    // 1. Read default retention config
+    final Map<DataHubRetentionKey, DataHubRetentionInfo> retentionPolicyMap =
+        parseFileOrDir(new ClassPathResource("./boot/retention.yaml").getFile());
+
+    // 2. Read plugin retention config files from input path and overlay
+    retentionPolicyMap.putAll(parseFileOrDir(new File(pluginPath)));
+
+    // 4. Set the specified retention policies
+    log.info("Setting {} policies", retentionPolicyMap.size());
+    boolean hasUpdate = false;
```

Comment:
why not just shortcircuit and call _retentionservice.batcha... inside the if directly?

Reply:
ah because, I want to batch apply after setting all retention policies. Each set retention sets a single policy and returns whether that set caused an actual update. If there is at least one update, we apply retention to all

Reply:
i see